src/BSON/UTCDateTime.c

```diff
@@ -47,17 +52,60 @@ PHONGO_API zend_class_entry *php_phongo_utcdatetime_ce;
 
 zend_object_handlers php_phongo_handler_utcdatetime;
 
-/* {{{ proto BSON\UTCDateTime UTCDateTime::__construct(integer $milliseconds)
+static void php_phongo_utcdatetime_init_from_current_time(php_phongo_utcdatetime_t *intern)
+{
+	int64_t        sec, usec;
+	struct timeval cur_time;
+
+	gettimeofday(&cur_time, NULL);
+	sec = cur_time.tv_sec;
+	usec = cur_time.tv_usec;
+
+	intern->milliseconds = (sec * 1000) + (usec / 1000);
+}
+
+static void php_phongo_utcdatetime_init_from_date(php_phongo_utcdatetime_t *intern, php_date_obj *datetime_obj)
+{
+	int64_t sec, usec;
+
+	/* The following assignments use the same logic as date_format() in php_date.c */
+	sec = datetime_obj->time->sse;
+	usec = (int64_t) floor(datetime_obj->time->f * 1000000 + 0.5);
+
+	intern->milliseconds = (sec * 1000) + (usec / 1000);
+}
+
+/* {{{ proto BSON\UTCDateTime UTCDateTime::__construct(integer|DateTime $milliseconds)
    Construct a new UTCDateTime */
 PHP_METHOD(UTCDateTime, __construct)
 {
 	php_phongo_utcdatetime_t    *intern;
-	zend_error_handling       error_handling;
-
+	zend_error_handling          error_handling;
+	zval                        *datetime = NULL;
 
 	zend_replace_error_handling(EH_THROW, phongo_exception_from_phongo_domain(PHONGO_ERROR_INVALID_ARGUMENT), &error_handling TSRMLS_CC);
 	intern = Z_UTCDATETIME_OBJ_P(getThis());
 
+#if PHP_VERSION_ID >= 50500
+	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS() TSRMLS_CC, "O", &datetime, php_date_get_immutable_ce()) == SUCCESS) {
+		php_phongo_utcdatetime_init_from_date(intern, Z_PHPDATE_P(datetime));
+
+		zend_restore_error_handling(&error_handling TSRMLS_CC);
+		return;
+	}
+#endif
+
+	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS() TSRMLS_CC, "|O!", &datetime, php_date_get_date_ce()) == SUCCESS) {
+		if (datetime) {
+			php_phongo_utcdatetime_init_from_date(intern, Z_PHPDATE_P(datetime));
+		} else {
+			php_phongo_utcdatetime_init_from_current_time(intern);
+		}
+
+		zend_restore_error_handling(&error_handling TSRMLS_CC);
+		return;
+	}
+
```

Comment:
I went with "Expected instance of DateTimeInterface, $classname given". We technically check for an instance of DateTime and DateTimeImmutable (on 5.5+); however, DateTimeInterface cannot be implemented by userland classes so the message is still accurate.